packages/core/src/sync-historical/index.ts

```diff
@@ -291,103 +293,305 @@ export const createHistoricalSync = async (
+  //   for (let i = 0; i < callTraces.length; i++) {
+  //     const callTrace = callTraces[i]!;
+  //     const block = blocks[i]!;
+
+  //     if (block.hash !== callTrace.blockHash) {
+  //       throw new Error(
+  //         `Detected inconsistent RPC responses. 'trace.blockHash' ${callTrace.blockHash} does not match 'block.hash' ${block.hash}`,
+  //       );
+  //     }
+
+  //     if (
+  //       block.transactions.find((t) => t.hash === callTrace.transactionHash) ===
+  //       undefined
+  //     ) {
+  //       throw new Error(
+  //         `Detected inconsistent RPC responses. 'trace.transactionHash' ${callTrace.transactionHash} not found in 'block.transactions' ${block.hash}`,
+  //       );
+  //     }
+  //   }
+
+  //   // Request transactionReceipts to check for reverted transactions.
+  //   const transactionReceipts = await Promise.all(
+  //     Array.from(transactionHashes).map((hash) =>
+  //       _eth_getTransactionReceipt(args.requestQueue, {
+  //         hash,
+  //       }),
+  //     ),
+  //   );
+
+  //   const revertedTransactions = new Set<Hash>();
+  //   for (const receipt of transactionReceipts) {
+  //     if (receipt.status === "0x0") {
+  //       revertedTransactions.add(receipt.transactionHash);
+  //     }
+  //   }
+
+  //   callTraces = callTraces.filter(
+  //     (trace) => revertedTransactions.has(trace.transactionHash) === false,
+  //   );
+
+  //   if (isKilled) return;
+
+  //   for (const hash of transactionHashes) {
+  //     if (revertedTransactions.has(hash) === false) {
+  //       transactionsCache.add(hash);
+  //     }
+  //   }
+
+  //   if (isKilled) return;
+
+  //   await args.syncStore.insertCallTraces({
+  //     callTraces: callTraces.map((callTrace, i) => ({
+  //       callTrace,
+  //       block: blocks[i]!,
+  //     })),
+  //     chainId: args.network.chainId,
+  //   });
+  // };
+
+  const syncTrace = async (
+    filter: TransactionFilter | TransferFilter,
+    blockNumber: number,
   ) => {
-    // Resolve `filter.toAddress`
-    let toAddress: Address[] | undefined;
-    if (isAddressFactory(filter.toAddress)) {
-      const childAddresses = await syncAddress(filter.toAddress, interval);
-      if (
-        childAddresses.length < args.common.options.factoryAddressCountThreshold
-      ) {
-        toAddress = childAddresses;
-      } else {
-        toAddress = undefined;
-      }
-    } else {
-      toAddress = filter.toAddress;
-    }
-
-    if (isKilled) return;
-
-    let callTraces = await _trace_filter(args.requestQueue, {
-      fromAddress: filter.fromAddress,
-      toAddress,
-      fromBlock: interval[0],
-      toBlock: interval[1],
-    }).then(
-      (traces) =>
-        traces.flat().filter((t) => t.type === "call") as SyncCallTrace[],
-    );
-
-    if (isKilled) return;
+    let traces = await _debug_traceBlockByNumber(args.requestQueue, {
+      blockNumber: toHex(blockNumber),
+    });
```

Comment:
When I was evaluating the debug rpc methods, I notice that `traceBlockByHash` was often way faster than `traceBlockByNumber`. It would require a different algorithm because block hash isn't immediately available, but something we should try before shipping